src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/Logging/OutputWindowLogger.cs

```diff
@@ -56,27 +56,54 @@ public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Except
         }
     }
 
-    private class OutputPane
+    private class OutputPane : IDisposable
     {
         private static readonly Guid s_outputPaneGuid = new("BBAFF416-4AF5-41F2-9F93-91F283E43C3B");
 
         private readonly JoinableTaskContext _threadingContext;
         private readonly IServiceProvider _serviceProvider;
+        private readonly AsyncQueue<string> _outputQueue;
+        private readonly CancellationTokenSource _disposalTokenSource;
         private IVsOutputWindowPane? _doNotAccessDirectlyOutputPane;
 
         public OutputPane(JoinableTaskContext threadingContext)
         {
             _threadingContext = threadingContext;
             _serviceProvider = ServiceProvider.GlobalProvider;
+
+            _outputQueue = new AsyncQueue<string>();
+            _disposalTokenSource = new CancellationTokenSource();
+
+            _ = StartListeningAsync();
+        }
+
+        private async Task StartListeningAsync()
+        {
+            // Ensure that we're never on the UI thread before we start listening, in case the async queue doesn't yield
+            // I suspect this is overkill :D
+            await TaskScheduler.Default.SwitchTo(alwaysYield: true);
+
+            while (!_disposalTokenSource.IsCancellationRequested)
+            {
+                await DequeueAsync(_disposalTokenSource.Token).ConfigureAwait(false);
+            }
         }
 
         public void WriteLine(string value)
         {
-            WriteLineInternal(value);
+            _outputQueue.Enqueue(value);
```

Comment:
```suggestion
            _ = _outputQueue.TryEnqueue(value);
```

This only fails if the queue is completed

Reply:
Oh good point. The last run of integration tests took 2 hours in debug, so I think one of these changes broke something, I wonder if this is it?